espressostreamer/utils.go

```diff
@@ -1,35 +1,46 @@
 package espressostreamer
 
+const (
+	FilterAndFind_Remove = iota
+	FilterAndFind_Keep
+	FilterAndFind_Target
+)
+
 // FilterAndFind filters an array in-place and returns the matching element based on a comparison function.
 // The comparison function should return:
-//   - 0 for the element to be returned
-//   - negative number for elements to be removed
-//   - positive number for elements to be kept
+//   - FilterAndFindTarget for the element to be returned, will be kept in the array
+//   - FilterAndFindKeep for elements to be kept
+//   - FilterAndFindRemove for elements to be removed
 //
-// Returns the found element (if any) and a boolean indicating if an element was found.
-func FilterAndFind[T any](arr *[]T, compareFunc func(T) int) (T, bool) {
-	var found T
+// Returns the index of the found element (if any)
+func FilterAndFind[T any](arr *[]T, compareFunc func(T) int) int {
+
 	var hasFound bool
+	idx := -1
 
 	if arr == nil || len(*arr) == 0 {
-		return found, false
+		return idx
 	}
 
 	j := 0
 	for i := 0; i < len(*arr); i++ {
 		result := compareFunc((*arr)[i])
+
+		if result == FilterAndFind_Remove || (result == FilterAndFind_Target && hasFound) {
+			continue
+		}
+
 		// Take the first element that matches
-		if result == 0 && !hasFound {
-			found = (*arr)[i]
+		if result == FilterAndFind_Target {
 			hasFound = true
-		} else if result > 0 {
-			if i != j {
-				(*arr)[j] = (*arr)[i]
-			}
-			j++
+			idx = j
+		}
+		if i != j {
+			(*arr)[j] = (*arr)[i]
```

Comment:
can you add some comments here? This code is confusing

Reply:
This function has comments like this:

```
// FilterAndFind filters an array in-place and returns the matching element based on a comparison function.
// The comparison function should return:
//   - FilterAndFindTarget for the element to be returned, will be kept in the array
//   - FilterAndFindKeep for elements to be kept
//   - FilterAndFindRemove for elements to be removed
```

So this is an `in-place` implementation. if `i` should be kept, then `j++`. If `i` should be removed, then continue.

`j` is the next legitimate index for the next legitimate `i`.

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/566/commits/16691fdbd6ca761b1c9e76ba292c6e6772aa3a2c